### DISCUSSION TITLE
Complex modify remaped function keys

### DISCUSSION THREAD

**Discussion by mietzen** [pqrs-org/Karabiner-Elements#3991] [Q&A] [answered]
Hey,

I'm struggling with remapping the volume keys e.g. `key_code: volume_decrement`

When using `key_code: volume_decrement` nothing happens, this seems to be a known issue:
https://github.com/pqrs-org/Karabiner-Elements/issues/3743
https://github.com/pqrs-org/Karabiner-Elements/issues/2735

I then used `key_code: f18` and it works 👍 

So I though I might use the `Function Keys` settings and remap the volume keys of my Apple A1243 to the F-Keys, this works, but now it doesn't trigger the complex modification anymore 🤣 
Is there a way to make this work?

![image](https://github.com/user-attachments/assets/63fc4c4c-67f9-4d5c-8f48-f76bacd3cc0c)

```
{
    "description": "Apple A1243 Sonos Volume Control",
    "manipulators": [
        {
            "conditions": [
                {
                    "identifiers": [
                        {
                            "description": "Apple A1243",
                            "product_id": 592,
                            "vendor_id": 1452
                        }
                    ],
                    "type": "device_if"
                }
            ],
            "from": { "key_code": "f17" },
            "to": [{ "shell_command": "sonos my-room mute $([[ $(sonos my-room mute) == 'off' ]] && echo 'on' || echo 'off')" }],
            "type": "basic"
        },
        {
            "conditions": [
                {
                    "identifiers": [
                        {
                            "description": "Apple A1243",
                            "product_id": 592,
                            "vendor_id": 1452
                        }
                    ],
                    "type": "device_if"
                }
            ],
            "from": { "key_code": "f18" },
            "to": [{ "shell_command": "sonos my-room relative_volume -5" }],
            "type": "basic"
        },
        {
            "conditions": [
                {
                    "identifiers": [
                        {
                            "description": "Apple A1243",
                            "product_id": 592,
                            "vendor_id": 1452
                        }
                    ],
                    "type": "device_if"
                }
            ],
            "from": { "key_code": "f19" },
            "to": [{ "shell_command": "sonos my-room relative_volume +5" }],
            "type": "basic"
        }
    ]
}
```

**Comment by tekezo** [comment, marked as answer]
If your goal is to change the f12 key (volume_decrement) behavior, you’ll need to modify the f12 key in Complex Modifications.

```json
{
    "description": "example",
    "manipulators": [
        {
            "from": {
                "key_code": "f12",
                "modifiers": { "optional": ["any"] }
            },
            "to": [{ "key_code": "spacebar" }],
            "type": "basic"
        }
    ]
}
```

**Reply by mietzen** [reply]
Thanks that worked! So is this a general rule?  Always take the physical `key_code` e.g. `f12` instead of the `consumer_key_code` like `volume_increment`? For complex modifications?

**Reply by tekezo** [reply]
> Always take the physical key_code e.g. f12 instead of the consumer_key_code like volume_increment?

Yes.
The events sent by the hardware are f1 - f12, so we need to specify these in Complex Modifications.

I'm planing to add a mode in Event Viewer to display the raw events from the hardware, which will make it easier to investigate the source events for modification.
